Mathlib/GroupTheory/Congruence/Basic.lean

```diff
@@ -69,6 +69,64 @@ protected def congr {c d : Con M} (h : c = d) : c.Quotient ≃* d.Quotient :=
   { Quotient.congr (Equiv.refl M) <| by apply Con.ext_iff.mp h with
     map_mul' := fun x y => by rcases x with ⟨⟩; rcases y with ⟨⟩; rfl }
 
+@[to_additive (attr := simp)]
+theorem congr_mk {c d : Con M} (h : c = d) (a : M) :
+    Con.congr h (a : c.Quotient) = (a : d.Quotient) := rfl
+
+@[to_additive]
+theorem le_comap_conGen {M N : Type*} [Mul M] [Mul N] (f : M → N)
+    (H : ∀ (x y : M), f (x * y) = f x * f y) (rel : N → N → Prop) :
+    conGen (fun x y ↦ rel (f x) (f y)) ≤ Con.comap f H (conGen rel) := by
+  intro x y h
+  simp only [Con.comap_rel]
+  exact .rec (fun x y h ↦ .of (f x) (f y) h) (fun x ↦ .refl (f x))
+    (fun _ h ↦ .symm h) (fun _ _ h1 h2 ↦ h1.trans h2) (fun {w x y z} _ _ h1 h2 ↦
+    (congrArg (fun a ↦ conGen rel a (f (x * z))) (H w y)).mpr
+    (((congrArg (fun a ↦ conGen rel (f w * f y) a) (H x z))).mpr
+    (.mul h1 h2))) h
+
+@[to_additive]
+theorem comap_conGen_of_bijective {M N : Type*} [Mul M] [Mul N] (f : M → N)
+    (hf : Function.Bijective f) (H : ∀ (x y : M), f (x * y) = f x * f y) (rel : N → N → Prop) :
```

Comment:
I feel like the proof below can be neater if we first prove it for a `MulEquiv` and then suit the hypotheses here using `MulEquiv.ofBijective`. Let me see if I can cook up something...

Reply:
Here's how I would use a `MulEquiv`. (Unfortunately I can't figure out how to make this suggestion...)

```lean
@[to_additive]
theorem comap_conGen_equiv {M N : Type*} [Mul M] [Mul N] (f : MulEquiv M N) (rel : N → N → Prop) :
    Con.comap f (map_mul f) (conGen rel) = conGen (fun x y ↦ rel (f x) (f y)) := by
  apply le_antisymm _ (le_comap_conGen f (map_mul f) rel)
  intro a b h
  simp only [Con.comap_rel] at h
  have H : ∀ n1 n2, (conGen rel) n1 n2 → ∀ a b, f a = n1 → f b = n2 →
      (conGen fun x y ↦ rel (f x) (f y)) a b := by
    intro n1 n2 h
    induction h with
    | of x y h =>
      intro _ _ fa fb
      apply ConGen.Rel.of
      rw [fa, fb]
      exact h
    | refl x =>
      intro _ _ fc fd
      rw [f.injective (fc.trans fd.symm)]
      exact ConGen.Rel.refl _
    | symm _ h => exact fun a b fs fb ↦ ConGen.Rel.symm (h b a fb fs)
    | trans _ _ ih ih1 =>
      exact fun a b fa fb ↦ Exists.casesOn (f.surjective _) fun c' hc' ↦
      ConGen.Rel.trans (ih a c' fa hc') (ih1 c' b hc' fb)
    | mul _ _ ih ih1 =>
      rename_i w x y z _ _
      intro a b fa fb
      rw [← f.eq_symm_apply, map_mul] at fa fb
      rw [fa, fb]
      exact ConGen.Rel.mul (ih (f.symm w) (f.symm x) (by simp) (by simp))
        (ih1 (f.symm y) (f.symm z) (by simp) (by simp))
  exact H (f a) (f b) h a b (refl _) (refl _)

@[to_additive]
theorem comap_conGen_of_bijective {M N : Type*} [Mul M] [Mul N] (f : M → N)
    (hf : Function.Bijective f) (H : ∀ (x y : M), f (x * y) = f x * f y) (rel : N → N → Prop) :
    Con.comap f H (conGen rel) = conGen (fun x y ↦ rel (f x) (f y)) :=
  comap_conGen_equiv (MulEquiv.ofBijective (MulHom.mk f H) hf) rel
```

The alternative would be to put the creation of a `MulEquiv` inline in the `mul` case of the induction; I think this is neater since it means less argument reshuffling for users who already have a `MulEquiv`.

Reply:
oh nice, thanks!